### PR TITLE
Bundled patches for #60, #61, #63 & #80

### PR DESCRIPTION
**Pull request by sjaeckel**
Please review:

@dmitry-lipetsk
@przemysx

### CODE REVIEW

File: bn_fast_s_mp_mul_digs.c [hunk 1]
```diff
@@ -87,7 +87,7 @@ int fast_s_mp_mul_digs(const mp_int *a, const mp_int *b, mp_int *c, int digs)
    {
       mp_digit *tmpc;
       tmpc = c->dp;
-      for (ix = 0; ix < (pa + 1); ix++) {
+      for (ix = 0; ix < pa; ix++) {
```
Comment (dmitry-lipetsk) [inline bn_fast_s_mp_mul_digs.c]: Looks ok.

File: bn_mp_lshd.c [hunk 2]
```diff
@@ -24,6 +24,10 @@ int mp_lshd(mp_int *a, int b)
    if (b <= 0) {
       return MP_OKAY;
    }
+   /* no need to shift 0 around */
+   if (mp_iszero(a) == MP_YES) {
+      return MP_OKAY;
+   }
```
Comment (dmitry-lipetsk) [inline bn_mp_lshd.c]: Looks ok.

**Comment by dmitry-lipetsk** [comment]
I vote to commit changes in
- bn_fast_s_mp_mul_digs.c
- bn_fast_s_mp_mul_digs.c

These problems are explicit and corrections are trivial.

The changes in:
- bn_fast_mp_montgomery_reduce.c
- bn_mp_montgomery_reduce.c

required rethink

**Comment by sjaeckel** [comment]
> The changes in:
> 
> * bn_fast_mp_montgomery_reduce.c
> * bn_mp_montgomery_reduce.c
>
> need rethink

1. the change in `bn_fast_mp_montgomery_reduce.c` checks that when `fast_mp_montgomery_reduce()` is called with a parameter that would overflow the array on the stack it errors out
2. the change in `bn_mp_montgomery_reduce.c` checks that it won't call `fast_mp_montgomery_reduce()` in case the check above would be triggered

**Comment by sjaeckel** [comment]
@dmitry-lipetsk do you need further explanations?

**Comment by dmitry-lipetsk** [comment]
Hello sjaeckel.

I already said my opinion :)

| The changes in:
|   bn_fast_mp_montgomery_reduce.c
|   bn_mp_montgomery_reduce.c
| need rethink

At current time I not has the time for this. Sorry.

**Comment by sjaeckel** [comment]
> At current time I not has the time for this. Sorry.

You're very welcome to re-open this issue after you had the time and the solution isn't okay.

**Pull request merged**